glean-core/ios/Glean/Metrics/EventMetric.swift

```
@@ -0,0 +1,224 @@
+    ///     * pingName: represents the name of the ping to retrieve the metric for.
+    ///                 Defaults to the first value in `sendInPings`.
+    /// - returns: true if metric value exists, otherwise false
+    func testHasValue(_ pingName: String? = nil) -> Bool {
+        Dispatchers.shared.assertInTestingMode()
+
+        let pingName = pingName ?? self.sendInPings[0]
+        return glean_event_test_has_value(Glean.shared.handle, self.handle, pingName) != 0
+    }
+
+    /// Deserializes an event in JSON into a RecordedEventData object.
+    ///
+    /// - parameters:
+    //      * jsonContent: The JSONObject containing the data for the event. It is in
+    ///       the same format as an event sent in a ping, and has the following entries:
+    ///         - timestamp (Int)
+    ///         - category (String): The category of the event metric
+    ///         - name (String): The name of the event metric
+    ///         - extra ([String: String]?): Map of extra key/value pairs
+    /// - returns: `RecordedEventData` representing the event data
+    private func deserializeEvent(_ jsonContent: [String: Any]) -> RecordedEventData? {
+        guard let category = jsonContent["category"] as? String else { return nil }
+        guard let name = jsonContent["name"] as? String else { return nil }
+        guard let timestamp = jsonContent["timestamp"] as? UInt64 else { return nil }
+        var extra: [String: String]?
+
+        if let extraObj = jsonContent["extra"] {
+            let extraObj = extraObj as? [String: String]
+            extra = extraObj
+        }
+
+        return RecordedEventData(category: category, name: name, timestamp: timestamp, extra: extra)
+    }
+
+    /// Returns the stored value for testing purposes only. This function will attempt to await the
+    /// last task (if any) writing to the the metric's storage engine before returning a value.
+    ///
+    /// Throws a `String` exception if no value is stored
```

nit: throws a `Missing value`

---

This also throws `Invalid JSON` and `Empty JSON`. I think this is throwing a lot of different exceptions. Do we really need that many exceptions communicated to the user? Maybe we can throw always the same exception (`Missing value`) and get away with Glean internal logging.

---

Good point, it's also just testing and invalid json is essentially just missing a value (like ... the only way this happens is memory corruption?)